### ISSUE TITLE
Cancellation of divergent recursive dataflows

### ISSUE THREAD

**Issue by vmarcos** [MaterializeInc/materialize#16800]
In PR #16787, it has been observed that if we render a divergent dataflow, it keeps on executing even if the corresponding view and indexes associated to it are dropped. We should think of a way to handle cancellation of dataflows that are non-terminating.

Here is an example of how to reproduce the issue:

````
CREATE VIEW v AS
WITH MUTUALLY RECURSIVE 
  cnt (cnt int) AS (
    SELECT  1 AS cnt
    UNION
    SELECT cnt+1 AS cnt FROM cnt)
SELECT * FROM cnt  WHERE cnt < 10;
CREATE DEFAULT INDEX ON v;
DROP VIEW v;
````

Now, in the memory visualizer, it is possible to see that the dataflow continues to be alive and producing ever increasing numbers of records.

**Comment by petrosagg** [comment]
That would be a great incentive to build confidence on [`Worker::drop_dataflow`](https://docs.rs/timely/latest/timely/worker/struct.Worker.html#method.drop_dataflow) and manage dataflows using dataflow-wide tokens which stop any dataflow regardless of where it's at in its execution!

I had put up a PR for this recently but we didn't want to potentially introduce further instability to the system because `drop_dataflow` isn't widely tested at the moment. https://github.com/MaterializeInc/materialize/pull/15779

**Comment by frankmcsherry** [comment]
A different guard rail we could use, and probably want anyhow, is to allow the user to bound the number of iterations. I'm guessing that debugging recursive queries is annoying, and a reasonable pattern is to look at the results after one iteration, two iterations, etc. You could change the query to time itself out (with some counter), but it's plenty easy for DD to handle this for you if you have an upper bound on number of iterations.

More generally .. though I'm sure also more frighteningly .. this could just be a temporal filter on round of iteration, which DD supports also (well, manipulation of the timestamp component, which is what temporal filters end up being).

**Comment by aalexandrov** [comment]
Summarizing some offline discussion with @antiguru.

It seems that we have two options for the frontend.

1. Extend the current WMR syntax with an optional `LIMIT x` clause (`WITH MUTUALLY RECURSIVE LIMIT $x`).
2. Introduce a query hint for the entire select query `OPTIONS (MAXRECURSION $x)`.
3. Introduce a session variable for all queries (`SET MAXRECURSION = $x`).

I'm leaning towards option (2) - the `OPTIONS` machinery already exists, and we can just forward the parameter when rendering the plan (that is, we don't need to extend `LetRec` to carry an additional parameter).

@frankmcsherry I'm interested in your opinion here.

**Comment by frankmcsherry** [comment]
The first option seems the best to me (although `LIMIT x` seems not great; it's unclear what it limits; records? iterations? minutes?). Not allowing different settings of the option for different WMR blocks will just bite you when you want to test a nested WMR.

I can't tell if the second option allows an `OPTIONS` for each `SELECT` command, or each `SELECT` clause. If the former it has the above problem, and if the latter there should be only one WMR possible in any case, and it seems easier to localize the argument to the clause it modifies.

**Comment by aalexandrov** [comment]
To summarize:

* We're leaning towards option (1).
* The proposed syntax is `WITH MUTUALLY RECURSIVE ITERATE $x TIMES`.

**Comment by ggevay** [comment]
I'll work now on option 1. from above, i.e., adding an optional clause to `WITH MUTUALLY RECURSIVE` to limit the number of iterations. As Frank noted above, this is also good for query debugging: when one wants to examine the output after a certain number of iterations.

This will reduce the problem of divergent dataflows, but it won't completely go away: Maybe the user doesn't specify this extra clause, and then she still ends up with a problematic dataflow that continues to be alive indefinitely. So we'll also want to work on properly cancelling the dataflows when the associated index and/or view is dropped.

**Comment by ggevay** [comment]
> a reasonable default value

Well, I'm not sure what value we would to set. The problem is that this value might be wildly different for different queries. A query that works with only small datasets might do thousands of iterations in a minute, but a query that is working with large datasets could run essentially forever even with tens of iterations. (Also, there are queries where even though the initial dataset is large and the first few iterations do a lot of work, there are still many iterations with small deltas.) So, I would be worried that having a non-infinite default value would bite some users, who don't realize that the default is not infinite, but at the same time wouldn't help some other users, whose queries would still run for a long time.

Btw. I'm not sure I fully understand why this problem is specific to WMR. Any query might take a very long time to complete, e.g., if a cross join surprisingly appears in a plan when a user forgets a join condition. Is the idea that in many of those situations, the query will OOM after a not very large amount of time anyway? Is this somehow less the case for WMR?

Another question: Does the problem come up only for VIEW dropping, or does it also happen with a SELECT? So, for a SELECT, does the dataflow go away if I hit Ctrl+C?

**Comment by aalexandrov** [comment]
> It seems to me SELECT also has the same problem.

Yes, but it seems we'll be prioritizing that work (#2392). @antiguru would you consider this as an alternative mechanism to handle divergent dataflows in case leave the "no limit option"?

**Comment by antiguru** [comment]
I'm not sure that #2392 is the right mechanism to prevent divergent dataflows. Recursive queries have memory requirements proportional to the trace, which means that we might OOM faster than the user has a chance to react. Limiting the number of recursions might help here.

> Another question: Does the problem come up only for VIEW dropping, or does it also happen with a SELECT? So, for a SELECT, does the dataflow go away if I hit Ctrl+C? Edit: It seems to me SELECT also has the same problem.

Currently, we let the computation finish, however long this might take. In the future, we might be able to uninstall the dataflows, but I'd prefer if we don't introduce this as a hard dependency.

I agree we have to think a bit more about this because we're mixing a product and an engineering question. I think it makes sense to have the infrastructure in place to set a limit on iterations and to have an optional global limit (or set the default to u64::MAX). We can then think about setting a default value and how to communicate it to users with product.

> Btw. it would be great to print a notice to the user when the limit is reached. Then we could set a relatively lower default. But I'm not sure whether we have the infrastructure for this.

Yep, this would be nice, but I don't think we have a way to communicate notices from the dataflow layer, only errors, but that's not what we'd like to have here.

**Comment by aalexandrov** [comment]
It seems that we are proposing various solution with two overlapping goals in mind:

1. Coming up with a mechanism and system behavior that protects customers from running divergent dataflows, especially on production clusters.
2. Coming up with mechanisms to inspect intermediate states of the dataflow.

We might decide that we use the same solution for both or use two separate solutions. Let's use this issue to track (1) and #18362  to track (2).

**Comment by ggevay** [comment]
Yes, we might want to have two limits: a soft limit for 2. and a hard limit that errors out for 1.

Note that 1. is easier, because it's ok if the limit setting is per-query (and/or a session variable), whereas for 2. it would be useful to have it per-WMR.

I'll implement 1. first, because that seems to be the most pressing matter. Then we can do 2. soon afterwards, possibly in two steps:
- First, we can have a version of 2. that is also per-query. This way, there is only a slight rendering change between 1. and 2., which should be easy. This already gives us the debugging benefit for non-nested WMR.
- The hard part for 2. is having it per-WMR: For this we need to somehow solve the problem [noted above](https://github.com/MaterializeInc/materialize/issues/16800#issuecomment-1480135678) that `NormalizeLets` sometimes merges two `LetRec` nodes into one, which is problematic if they have differing limits.

**Comment by aalexandrov** [comment]
[I added a comment in #18362](https://github.com/MaterializeInc/materialize/issues/18362#issuecomment-1481605865) with a straight-forward way to emulate (2) in SQL using mechanic changes to an existing WMR query. It could be that this solution is more robust w.r.t. the `NormalizeLets` concerns that you raised above. In any case it should unblock people that want to debug #18362 until we support the more elegant direct syntax.

**Comment by aalexandrov** [comment]
@ggevay can you please fix the TODO(#16800) in `column_knowledge.rs` that is added in #18323 when you open a PR for the hard limit?

**Comment by ggevay** [comment]
Yes, will do, thanks for the reminder!

**Comment by philip-stoev** [comment]
@teskje 's https://github.com/MaterializeInc/materialize/pull/18442 is effective in cancelling divergent WMR dataflows.

**Comment by petrosagg** [comment]
> One of those will be shutting down WMR dataflows by inserting an operator in each loop that stops producing output when it sees a shutdown token being dropped

I wanted to propose the same thing, it sounds like the easiest thing to do with our existing token infrastructure. The implementation was only a few lines of code so I pushed it here https://github.com/MaterializeInc/materialize/pull/18718

**Comment by teskje** [comment]
#18718 is merged now, so divergent WMR dataflows are cancelled as one would expect. @ggevay @aalexandrov I suggest we close this issue here as complete. If we still want a hard limit (we might not?) we should open a separate ticket to track that specifically.

**Comment by ggevay** [comment]
Thanks a lot @teskje and @petrosagg!

We'll discuss the hard limit later when @antiguru comes back. (I think he mentioned it to me once that a default hard limit might be desirable even if we have dataflow cancellation.) I opened an issue: https://github.com/MaterializeInc/materialize/issues/18832.
